### ISSUE TITLE
Output celebrations in a array by date

### ISSUE THREAD

**Issue by tukusejssirs** [pejulian/romcal#270]
It might be too late for v2.0.0 to change it, but I think we should think wnd talk about this for v3.0.0.

Currently, we output every celebration as a separate (let’s say) object.

I propose we should create an array of sequential dates limited by start end time (like start/end of a year). Each date would be an array itself and would contain an array of celebrations.

This way, the main array (array of dates) for civil years would contain either 365 or 366 elements. Also we would eliminate the duplicate definition of dates for days where there are two or more celebration options.

**Comment by emagnier** [comment]
By default, romcal output exactly 365 (or 366) date items for a civil year, so indeed, only one item per date.

When we need to output all optional memorials, so potentially multiple date items per dates, we need to set `outputOptionalMemorials` to `true` (this option is `false` by default).

I'm not really convinced to change the default returned array to an object of dates. Because on a lot of use cases, like filtering or querying specific criteria, the developer will have to convert it back to an array. And in terms of performances this isn't ideal (on this specific topic, we already have a lot on our plate some).

By the way, on romcal v2, there is already methods to address your specific wish:
https://github.com/romcal/romcal/blob/_F-TypeScript/docs/general-usage.md#-group-by-a-predefined-criteria-groupbycriteria

But generally, I will instead encourage the use of the [`.reduce`](https://github.com/romcal/romcal/blob/_F-TypeScript/docs/general-usage.md#-group-by-any-criteria) method in JavaScript. This is now natively supported everywhere in this coding language, and will allow addressing all specific needs a developer can have for its development requirements.
This is for the same reasons why I've redesigned the [filtering](https://github.com/romcal/romcal/blob/_F-TypeScript/docs/general-usage.md#filter-by-date) options from romcal < v2, since it's already possible to do it natively (and in a simpler and explicit way). I only kept the `.groupBy` method for amateur developers (because it might be a bit more difficult for them, and to provide syntactic sugar since it take a bit more code).

**Comment by tukusejssirs** [comment]
> By default, romcal output exactly 365 (or 366) date items for a civil year, so indeed, only one item per date.
> 
> When we need to output all optional memorials, so potentially multiple date items per dates, we need to set `outputOptionalMemorials` to `true` (this option is `false` by default).

As for the default value of `outputOptionalMemorials`, we discussed it in #250 and in [one of your comments](https://github.com/romcal/romcal/pull/250#issuecomment-735299623) you said _I'm not opposed to reverse this option. This could be replaced by a "strict mode" that will only output one item per day, but `false` by default._ I understood this that you then decided to change the default option to `true` (i.e. output optional memorials by default) which is what romcal does in pre-2.0.0 versions and what I wish romcal would do.

Even if `outputOptionalMemorials` would be `false` by default, I think it would be useful to implement this issue (#270). While it won’t be any useful when `outputOptionalMemorials` is `false`, but when it is `true`, one could easily and more performantly check the number of all possible celebrations on a particular day using `romcal['2021-04-06'].length` (where `romcal` is the object output by romcal).

> I'm not really convinced to change the default returned array to an object of dates. Because on a lot of use cases, like filtering or querying specific criteria, the developer will have to convert it back to an array. And in terms of performances this isn't ideal (on this specific topic, we already have a lot on our plate some).

Why would a developer need to convert it back to an array? If they know the structure of the object output by romcal, it should not be needed to restructure the object.

> By the way, on romcal v2, there is already methods to address your specific wish:
> https://github.com/romcal/romcal/blob/_F-TypeScript/docs/general-usage.md#-group-by-a-predefined-criteria-groupbycriteria
> 
> But generally, I will instead encourage the use of the [`.reduce`](https://github.com/romcal/romcal/blob/_F-TypeScript/docs/general-usage.md#-group-by-any-criteria) method in JavaScript. This is now natively supported everywhere in this coding language, and will allow addressing all specific needs a developer can have for its development requirements.
> This is for the same reasons why I've redesigned the [filtering](https://github.com/romcal/romcal/blob/_F-TypeScript/docs/general-usage.md#filter-by-date) options from romcal < v2, since it's already possible to do it natively (and in a simpler and explicit way). I only kept the `.groupBy` method for amateur developers (because it might be a bit more difficult for them, and to provide syntactic sugar since it take a bit more code).

Okay, I see it is better to use the `.reduce` methode in comparison with `groupBy` method and also with the proposal of this issue.

---

Just to be clear: the closure of this issue is accepted! :wink:

**Comment by emagnier** [comment]
Thank you for reminding me of this [previous discussion](https://github.com/romcal/romcal/pull/250#issuecomment-735299623). I knew we had talked about something close to it, but couldn't remember what & where. But in fact this isn't exactly the same things, and on this other topic, I'm still in favor of output all possible liturgical days by default (and enable a strict mode from an option).

And as I wrote above, the things you are asking [already exist](https://github.com/romcal/romcal/blob/_F-TypeScript/docs/general-usage.md#-group-by-a-predefined-criteria-groupbycriteria). 
(this is why I'm removing the "wontfix" label on this issue)

> Why would a developer need to convert it back to an array? If they know the structure of the object output by romcal, it should not be needed to restructure the object.

The only case where you don't need to convert it back to an array, is when you know exactly the date (if liturgical days are grouped by date, as you are asking). However a `.getLiturgicalDay` method already exists for this need (it's a sugar syntactic method that filter an array internally). But for any other criteria, you will need to refine your results from an array.
